tutorials/decentralized-exchange-ink/1-implementation.mdx

```diff
@@ -0,0 +1,152 @@
+        total_shares: Balance, // Stores the total amount of share issued for the pool
+        shares: BTreeMap<AccountId, Balance>, // Stores the share holding of each user
+    }
+```
+
+## Constructor
+
+Constructor is the hear of every smart contract. As a parameter to constructor our smart contract we pass information about fees.
+
+```rust
+/// Instantiating AMM instance
+/// @param _fees: valid interval -> [0,1000)
+#[ink(constructor)]
+pub fn new(_fees: Balance) -> Self {
+    Self {
+        trading_fee: if _fees >= 1000 { 0 } else { _fees },
+        ..Default::default()
+    }
+}
+```
+
+## Business logic and core functions
+
+The core functions of every decentralized exchange are:
+
+- deliver liquidity
+- swap tokens
+- withdraw liquidity
+
+Let's start from crucial subject in any decentralized exchange which is liquidity, below we can function responsible for providing liquidity:
+
+```rust
+#[ink(message)]
+pub fn provide_liquidity(
+    &mut self,
+    _amount_token1: Balance,
+    _amount_token2: Balance,
+) -> Result<Balance, Error>
+```
+
+`provide_liquidity` in `ink!` smart contract programming language every public function has to have `#[ink(message)]` attribute. There need to be at least one public method in a smart contract.
+
+`swap` feature is the most complex one in the smart contract and due to a need for estimation firstly of possibility a given swap it is implemented two functions `estimate_swap_token1_for_given_token1` and `estimate_swap_token1_for_given_token2`.
+
+`withdraw` function is quite intuitive where we check firstly validation of the process with `check_valid_amount` to be sure that some user is not able to withdraw more money that it is previously delivered.
+
+## Errors definitions and utils
+
+Rust does not have traditional exceptions like those found in languages such as Java or Python. Instead, Rust relies on the `Result` and `Option` types for error handling and handling absence of values, respectively.
+
+```rust
+    /// Errors definitions
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        PoolDepleted(String),
+        ZeroAmountErr(String),
+        InvalidShareErr(String),
+        ZeroLiquidityErr(String),
+        SlippageExceededErr(String),
+        InsufficientAmountErr(String),
+        NonEquivalentValueErr(String),
+        ThresholdNotReachedErr(String),
+        InsufficientLiquidityErr(String),
+    }
+```
+
+In the smart contract is implement some functions that are not strictly needed but helpful for testing, for example a place where a software engineer is able to generate tokens for tests - `faucet_brrr`. To follow K curve in an automated market maker is implemented `get_k` function. To get information about portfolio is implemented `get_information_portfolio` function and for pools information `get_pool_details`.
```

Comment:
Same as above. Give some reference to the functions you are quoting here. Either links to lines in the github repo or better, copy the relevant parts to this tutorial and put it in cotext. At the moment I am a little lost on what you are talking about.

Reply:
agree, hope now is better